commit 220b1073841110710cc1c68730b04f5b8a2ad763
Author: ish <ish@innogrid.com>
Date:   Fri Sep 13 10:54:57 2024 +0900

    Azure: Add json tags to MonitoringHandler's data

diff --git a/cloud-control-manager/cloud-driver/interfaces/resources/MonitoringHandler.go b/cloud-control-manager/cloud-driver/interfaces/resources/MonitoringHandler.go
index 54f88e7f..e32a3061 100644
--- a/cloud-control-manager/cloud-driver/interfaces/resources/MonitoringHandler.go
+++ b/cloud-control-manager/cloud-driver/interfaces/resources/MonitoringHandler.go
@@ -32,14 +32,14 @@ type ClusterNodeMonitoringReqInfo struct {
 }
 
 type TimestampValue struct {
-	Timestamp time.Time
-	Value     string
+	Timestamp time.Time `json:"timestamp"`
+	Value     string    `json:"value"`
 }
 
 type MetricData struct {
-	MetricName      string
-	MetricUnit      string
-	TimestampValues []TimestampValue
+	MetricName      string           `json:"metricName"`
+	MetricUnit      string           `json:"metricUnit"`
+	TimestampValues []TimestampValue `json:"timestampValues"`
 }
 
 type MetricType string
